### PR TITLE
in-place ops to alias tensor args of function

### PR DESCRIPTION
**Pull request by crcrpar**
## What does this PR do?

This is based on #798 and #869 and is trying to manage to support #738.

cc @borda @apaz-cli

### CODE REVIEW

File: thunder/core/proxies.py [hunk 1]
```diff
@@ -1621,6 +1639,11 @@ def tensorproxy(t: torch.Tensor, /, *, name: None | str, history: None | tuple =
     # See Note [DistributedDataParallel and distparallel_type]
     distparallel_type = getattr(t, "distparallel_type", None)
     _thunder_fsdp_padding_size = getattr(t, "_thunder_fsdp_padding_size", None)
+    _storage_data_ptr = (
+        t.untyped_storage().data_ptr()
+        if torch.is_tensor(t) and not t.is_sparse
```
Comment (crcrpar) [inline thunder/core/proxies.py:1644]:
```suggestion
        if torch.is_tensor(t) and t.layout == pytorch.strided
```

File: thunder/core/transform_common.py [hunk 2]
```diff
@@ -569,7 +641,9 @@ def create_functional_bsym_from(inplace_bsym: BoundSymbol) -> BoundSymbol:
 
 
 def functionalize_inplace_ops(
-    computation_trace: Trace, orig_to_view_swap_map: dict[VariableInterface, TensorProxy]
+    computation_trace: Trace,
```
Comment (crcrpar) [inline thunder/core/transform_common.py:644]: this needs refactoring. It'd not be good to have a function of about 400 lines

**Review by t-vi — APPROVED** [review]
Seems OK, thank you @crcrpar

File: thunder/__init__.py [hunk 3]
```diff
@@ -550,11 +564,16 @@ def get_computation_and_inputs(*args, **kwargs):
...
-                        for bsym in filter(lambda b: b.sym.id == prims.PrimIDs.COPY_, computation_trc.bound_symbols)
-                        if bsym.flat_proxy_args[1] in arg_to_idx
-                    }
+                    tensor_indices: int = []
+                    tensor_args_consumed_by_inplace_grouped_by_numel: dict[int, list[TensorProxy]] = defaultdict(list)
+                    for bsym in filter(lambda b: b.sym.id == prims.PrimIDs.COPY_, computation_trc.bound_symbols):
+                        t = bsym.flat_proxy_args[1]
+                        index = arg_to_idx[t]
+                        numel = t.numel
+                        tensor_args_consumed_by_inplace_grouped_by_numel[numel].append(index)
```
Comment (jjsjann123) [inline thunder/__init__.py:573]: question here: what's the reason that we are grouping them by numel?

File: thunder/__init__.py [hunk 4]
```diff
@@ -550,11 +564,16 @@ def get_computation_and_inputs(*args, **kwargs):
...
+                        tensor_indices.append(index)
+                    if len(tensor_args_consumed_by_inplace_grouped_by_numel) > 1:
```
Comment (jjsjann123) [inline thunder/__init__.py:575]: follow up question, what's the implication of the bucket size > 1?

**Pull request merged**